runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -2001,25 +2001,7 @@ TR::Register *J9::X86::TreeEvaluator::evaluateNULLCHKWithPossibleResolve(
    //
    if (needLateEvaluation)
       {
-      if (comp->useCompressedPointers())
-         {
-         // for stores under NULLCHKs, artificially bump
-         // down the reference count before evaluation (since stores
-         // return null as registers)
-         //
-         bool fixRefCount = false;
-         if (node->getFirstChild()->getOpCode().isStoreIndirect() &&
-               node->getFirstChild()->getReferenceCount() > 1)
-            {
-            node->getFirstChild()->decReferenceCount();
-            fixRefCount = true;
-            }
-         cg->evaluate(node->getFirstChild());
-         if (fixRefCount)
-            node->getFirstChild()->incReferenceCount();
-         }
-      else
-         cg->evaluate(firstChild);
+      cg->evaluate(firstChild);
```

Comment:
Found the issue. It appears that the code is slightly different on x86 than on Z and Power in this area. The mistake here is the new line we added should be `cg->evaluate(node->getFirstChild());`. This is because on x86 we modify the value of `firstChild` where as on Z and Power we do not.

Here is explicitly where we do this in the code on x86:
https://github.com/eclipse/openj9/blob/137829daa49943b5129f54bdac8aa5754b2e3186/runtime/compiler/x/codegen/J9TreeEvaluator.cpp#L1665-L1688

And for reference here it is on Z for example (where we use a separate variable `n` to do this):
https://github.com/eclipse/openj9/blob/137829daa49943b5129f54bdac8aa5754b2e3186/runtime/compiler/z/codegen/J9TreeEvaluator.cpp#L5571-L5590

And here it is on Power:
https://github.com/eclipse/openj9/blob/137829daa49943b5129f54bdac8aa5754b2e3186/runtime/compiler/p/codegen/J9TreeEvaluator.cpp#L13111-L13124

Reply:
In fact I believe x86 has a bug here, or at the very least we are missing opportunities. Because we modify the value of `firstChild` the computation on line 2569:
https://github.com/eclipse/openj9/blob/137829daa49943b5129f54bdac8aa5754b2e3186/runtime/compiler/x/codegen/J9TreeEvaluator.cpp#L1659-L1713

i.e. this line:
https://github.com/eclipse/openj9/blob/137829daa49943b5129f54bdac8aa5754b2e3186/runtime/compiler/x/codegen/J9TreeEvaluator.cpp#L1690

Means that all the subsequent `if` and `else if` checks will fail! This is because `opcode` is an indirect load IL, and all the checks are checking for something else, for example:

```
   if (opCode.isLoadVar() || (comp->target().is64Bit() && opCode.getOpCodeValue()==TR::l2i))
...
   else if (opCode.isStore())
...
   else if (opCode.isCall()     &&
            opCode.isIndirect() &&
            cg->getNumberBytesReadInaccessible() > TR::Compiler->om.offsetOfObjectVftField())
...
   else if (opCode.getOpCodeValue() == TR::monent ||
            opCode.getOpCodeValue() == TR::monexit)
...
   else if (!disableBranchlessPassThroughNULLCHK && opCode.getOpCodeValue () == TR::PassThrough
            && !needResolution && cg->getHasResumableTrapHandler())
...
```

All of these checks will fail. Surely this is a bug right?

Reply:
@0xdaryl I think this may be outside the scope of this PR to fix because the above code has been dead for quite a while. I'm not sure what will happen if we suddenly start running all that code. Let me know what you think on how we should proceed here. The way I see it we have a few options:

1. Leave the code as is for now and open up an issue
2. Try enabling the code and see if any tests fail
3. Wait until we can common up this evaluator as much of it can be shared across platforms and fix it then
4. ???

Reply:
Thanks for the thorough scouring of this code.

I think the output of the loop on L1683-L1684 will either be an indirect load opcode or a readbar opcode.  If it is an indirect load opcode then the condition on L1710 will still succeed because an indirect load will answer `true` for an `isLoadVar()` query.  `readbar`s do seem to skip this logic and will generate an explicit NULLCHK (it isn't clear to me if that is the expected behaviour or not).

Can you push a fix for just the problem found in this PR?  Any further work we need to investigate can be handled in a separate issue.